### PR TITLE
feat(ceremony verification): implemented ceremony finalization verification 

### PR DESCRIPTION
**Pull request by ctrlc03**
Implemented a function that given a ceremony prefix will verify whether the finalization was successful: all artifacts are uploaded and are valid. These are compared with local copies that are generated on the fly. Ceremony verification is now a Hardhat task as it requires smart contract interaction.

- [x] look into Functions deployment failure. What changed that could have caused the [error](https://github.com/quadratic-funding/mpc-phase2-suite/pull/348#issuecomment-1468486286).  Is it related to rollup?
- [x] Take WASM from S3
- [x] Take circuit inputs from file
- [x] Upgrade Cloud Functions memory limit
- [x] Run as cli command using hardhat task

Closes:
* https://github.com/privacy-scaling-explorations/groth-p0tion/issues/13
* https://github.com/privacy-scaling-explorations/groth-p0tion/issues/56
* https://github.com/privacy-scaling-explorations/groth-p0tion/issues/68

### CODE REVIEW

**Comment by gurrpi** [comment]
Couldn't deploy Firebase functions:

```
% firebase deploy --only functions --project p0tion-dev-branch

=== Deploying to 'p0tion-dev-branch'...

i  deploying functions
Running command: yarn --prefix "$RESOURCE_DIR" build
yarn run v1.22.19
$ rimraf dist && rollup -c rollup.config.ts --configPlugin typescript

src/functions/index.ts → ./dist/src/functions/index.node.js, ./dist/src/functions/index.node.mjs...
(!) Unresolved dependencies
https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency
firebase/functions (imported by "../actions/src/helpers/functions.ts" and "../actions/src/helpers/services.ts")
mime-types (imported by "../actions/src/helpers/storage.ts" and "src/lib/utils.ts")
firebase/firestore (imported by "../actions/src/helpers/database.ts", "../actions/src/helpers/verification.ts" and "../actions/src/helpers/services.ts")
ffjavascript (imported by "../actions/src/helpers/utils.ts" and "../actions/src/helpers/contracts.ts")
firebase/app (imported by "../actions/src/helpers/services.ts")
firebase/auth (imported by "../actions/src/helpers/authentication.ts")
ethers (imported by "../actions/src/helpers/contracts.ts")
solc (imported by "../actions/src/helpers/contracts.ts")
firebase-admin/firestore (imported by "src/lib/utils.ts" and "src/functions/circuit.ts")
node:fs (imported by "src/lib/utils.ts")
node:stream (imported by "src/lib/utils.ts")
node:util (imported by "src/lib/utils.ts")
timers/promises (imported by "src/lib/utils.ts")
firebase-functions/v1 (imported by "src/functions/circuit.ts")
firebase-functions/v2 (imported by "src/functions/circuit.ts")
(!) "this" has been rewritten to "undefined"
https://rollupjs.org/guide/en/#error-this-is-undefined
../../node_modules/solc/index.js
1: "use strict";
2: var __importDefault = (this && this.__importDefault) || function (mod) {
                          ^
3:     return (mod && mod.__esModule) ? mod : { "default": mod };
4: };
...and 1 other occurrence
created ./dist/src/functions/index.node.js, ./dist/src/functions/index.node.mjs in 1.5s
✨  Done in 4.25s.
✔  functions: Finished running predeploy script.
i  functions: preparing codebase default for deployment
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
i  artifactregistry: ensuring required API artifactregistry.googleapis.com is enabled...
✔  functions: required API cloudbuild.googleapis.com is enabled
✔  functions: required API cloudfunctions.googleapis.com is enabled
✔  artifactregistry: required API artifactregistry.googleapis.com is enabled
⚠  functions: package.json indicates an outdated version of firebase-functions. Please upgrade using npm install --save firebase-functions@latest in your functions directory.
i  functions: Loaded environment variables from .env.
i  functions: preparing . directory for uploading...
i  functions: packaged /Users/daehyun/workspace/mpc-phase2-suite/packages/backend (655.8 KB) for uploading
i  functions: packaged /Users/daehyun/workspace/mpc-phase2-suite/packages/backend (656.01 KB) for uploading
i  functions: ensuring required API cloudscheduler.googleapis.com is enabled...
✔  functions: required API cloudscheduler.googleapis.com is enabled
i  functions: ensuring required API run.googleapis.com is enabled...
i  functions: ensuring required API eventarc.googleapis.com is enabled...
i  functions: ensuring required API pubsub.googleapis.com is enabled...
i  functions: ensuring required API storage.googleapis.com is enabled...
✔  functions: required API storage.googleapis.com is enabled
✔  functions: required API eventarc.googleapis.com is enabled
✔  functions: required API run.googleapis.com is enabled
✔  functions: required API pubsub.googleapis.com is enabled
i  functions: generating the service identity for pubsub.googleapis.com...
i  functions: generating the service identity for eventarc.googleapis.com...
✔  functions: . folder uploaded successfully
i  functions: Skipping the deploy of unchanged functions with experimental support for skipdeployingnoopfunctions
i  functions: updating Node.js 16 function checkAndPrepareCoordinatorForFinalization(us-central1)...
i  functions: updating Node.js 16 function checkAndRemoveBlockingContributor(us-central1)...
i  functions: updating Node.js 16 function checkIfObjectExist(us-central1)...
i  functions: updating Node.js 16 function checkParticipantForCeremony(us-central1)...
i  functions: updating Node.js 16 function completeMultiPartUpload(us-central1)...
i  functions: updating Node.js 16 function coordinateCeremonyParticipant(us-central1)...
i  functions: updating Node.js 16 function createBucket(us-central1)...
i  functions: updating Node.js 16 function finalizeCeremony(us-central1)...
i  functions: updating Node.js 16 function finalizeCircuit(us-central1)...
i  functions: updating Node.js 16 function generateGetObjectPreSignedUrl(us-central1)...
i  functions: updating Node.js 16 function generatePreSignedUrlsParts(us-central1)...
i  functions: updating Node.js 16 function initEmptyWaitingQueueForCircuit(us-central1)...
i  functions: updating Node.js 16 function permanentlyStoreCurrentContributionTimeAndHash(us-central1)...
i  functions: updating Node.js 16 function processSignUpWithCustomClaims(us-central1)...
i  functions: updating Node.js 16 function progressToNextCircuitForContribution(us-central1)...
i  functions: updating Node.js 16 function progressToNextContributionStep(us-central1)...
i  functions: updating Node.js 16 function refreshParticipantAfterContributionVerification(us-central1)...
i  functions: updating Node.js 16 function registerAuthUser(us-central1)...
i  functions: updating Node.js 16 function resumeContributionAfterTimeoutExpiration(us-central1)...
i  functions: updating Node.js 16 function setupCeremony(us-central1)...
i  functions: updating Node.js 16 function startCeremony(us-central1)...
i  functions: updating Node.js 16 function startMultiPartUpload(us-central1)...
i  functions: updating Node.js 16 function stopCeremony(us-central1)...
i  functions: updating Node.js 16 function temporaryStoreCurrentContributionMultiPartUploadId(us-central1)...
i  functions: updating Node.js 16 function temporaryStoreCurrentContributionUploadedChunkData(us-central1)...
✔  functions[verifycontribution(us-central1)] Skipped (No changes detected)
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.

Functions deploy had errors with the following functions:
        checkAndPrepareCoordinatorForFinalization(us-central1)
        checkAndRemoveBlockingContributor(us-central1)
        checkIfObjectExist(us-central1)
        checkParticipantForCeremony(us-central1)
        completeMultiPartUpload(us-central1)
        coordinateCeremonyParticipant(us-central1)
        createBucket(us-central1)
        finalizeCeremony(us-central1)
        finalizeCircuit(us-central1)
        generateGetObjectPreSignedUrl(us-central1)
        generatePreSignedUrlsParts(us-central1)
        initEmptyWaitingQueueForCircuit(us-central1)
        permanentlyStoreCurrentContributionTimeAndHash(us-central1)
        processSignUpWithCustomClaims(us-central1)
        progressToNextCircuitForContribution(us-central1)
        progressToNextContributionStep(us-central1)
        refreshParticipantAfterContributionVerification(us-central1)
        registerAuthUser(us-central1)
        resumeContributionAfterTimeoutExpiration(us-central1)
        setupCeremony(us-central1)
        startCeremony(us-central1)
        startMultiPartUpload(us-central1)
        stopCeremony(us-central1)
        temporaryStoreCurrentContributionMultiPartUploadId(us-central1)
        temporaryStoreCurrentContributionUploadedChunkData(us-central1)
i  functions: cleaning up build files...
Error: There was an error deploying functions:
- Error Failed to update function resumeContributionAfterTimeoutExpiration in region us-central1
- Error Failed to update function permanentlyStoreCurrentContributionTimeAndHash in region us-central1
- Error Failed to update function generateGetObjectPreSignedUrl in region us-central1
- Error Failed to update function progressToNextContributionStep in region us-central1
- Error Failed to update function progressToNextCircuitForContribution in region us-central1
- Error Failed to update function temporaryStoreCurrentContributionMultiPartUploadId in region us-central1
- Error Failed to update function generatePreSignedUrlsParts in region us-central1
- Error Failed to update function checkIfObjectExist in region us-central1
- Error Failed to update function createBucket in region us-central1
- Error Failed to update function checkAndRemoveBlockingContributor in region us-central1
- Error Failed to update function checkAndPrepareCoordinatorForFinalization in region us-central1
- Error Failed to update function processSignUpWithCustomClaims in region us-central1
- Error Failed to update function stopCeremony in region us-central1
- Error Failed to update function refreshParticipantAfterContributionVerification in region us-central1
- Error Failed to update function initEmptyWaitingQueueForCircuit in region us-central1
- Error Failed to update function coordinateCeremonyParticipant in region us-central1
- Error Failed to update function temporaryStoreCurrentContributionUploadedChunkData in region us-central1
- Error Failed to update function startCeremony in region us-central1
- Error Failed to update function finalizeCeremony in region us-central1
- Error Failed to update function startMultiPartUpload in region us-central1
- Error Failed to update function completeMultiPartUpload in region us-central1
- Error Failed to update function finalizeCircuit in region us-central1
- Error Failed to update function checkParticipantForCeremony in region us-central1
- Error Failed to update function setupCeremony in region us-central1
- Error Failed to update function registerAuthUser in region us-central1
```

**Comment by 0xjei** [comment]
@ctrlc03 @gurrpi I can handle this quickly. I will add a review after verifying the solution locally.

**Review by 0xjei — CHANGES_REQUESTED** [review]
Left some comments to fix the Cloud Function deploy issue!

Anyway, according to what the GCP Log Explorer pointed out, in addition to `solc`, `ethers` would also be missing in the `backend` package. 

Modifying the types generated with `dts-gen` of `solc` will also require a code-side modification of the `solc.compile()` method call. Previously, the old types interface was covering the mandatory request for the second parameter ("p1") that involves imports for the contract. You can create a method similar to the one in the official documentation (https://github.com/ethereum/solc-js#example-usage-with-import-callback).

After you solve these steps, Cloud Functions could return to you this error `Function cannot be initialized. Error: memory limit exceeded.`, except for `verifycontribution` which has higher memory than the default (16GiB vs default 256MiB). I would say that there is absolutely no huge impact on costs by upgrading the functions to `512MiB` as deducted from the estimates conducted with the spreadsheet. Also because this is an upper limit, the functions could potentially use the memory but do not work 24/7 at max memory rate.

To recap, the steps should be as follows.
1. A) Install `solc` and `ethers` inside the `backend` package.
1. B) Install `solc` inside the `phase2cli` package.
2) Update the `solc` types as described in the comments (use `dts-gen`) and add the `solc.d.ts` to each package in the relevant `types` folder
3) Add the function for retrieving imports and call `solc.compile()` passing that function as the second parameter.
4) Update each cloud function in the `backend` package by incrementing the memory simply by adding `runWith({ memory: "512MB"})` in the definition (example below).

```js
export const checkParticipantForCeremony = functions.runWith({
    memory: "512MB"
}).https.onCall( ... 
```
5) Run this command from the root `yarn && yarn build && cd packages/backend && yarn firebase:deploy` 
6) Verify from the GCP Cloud Functions logs that everything is working.

Hope this solve the issue and unblock the next verification steps, lmk ^^

File: packages/actions/package.json [hunk 1]
```diff
@@ -36,7 +36,8 @@
         "build:watch": "rollup -c rollup.config.ts -w --configPlugin typescript",
         "pre:publish": "yarn build",
         "compile:contracts": "hardhat compile",
-        "test:contracts": "hardhat test test/unit/contract.test.ts"
+        "test:contracts": "export GOOGLE_APPLICATION_CREDENTIALS=\"../backend/serviceAccountKey.json\" && NODE_ENV=prod hardhat test test/unit/contract.test.ts",
+        "verify:ceremony": "export GOOGLE_APPLICATION_CREDENTIALS=\"../backend/serviceAccountKey.json\" && hardhat verifyCeremony"
```
Comment (0xjei) [inline packages/actions/package.json:40]: `NODE_ENV=prod` is missing here

File: packages/actions/src/helpers/contracts.ts [hunk 2]
```diff
@@ -58,3 +74,176 @@ export const verifyGROTH16ProofOnChain = async (contract: any, proof: any): Prom
...
+ * @param functions <Functions> firebase functions instance
+ * @param firestore <Firestore> firebase firestore instance
+ * @param ceremonyPrefix <string> ceremony prefix
+ * @param outputDirectory <string> output directory where to store the ceremony artifacts
+ * @param solidityVersion <string> solidity version to use for the verifier contract
+ * @param wasmPath <string> path to the wasm file
+ * @param circuitInputs <object> circuit inputs
+ * @param logger <any> logger for printing snarkjs output
+ */
+export const verifyCeremony = async (
```
[the remark this answers is not in the document]
Comment (0xjei) [reply packages/actions/src/helpers/contracts.ts:144]: However, making these changes will solve the following issues:
- requesting information that is tied to individual circuits when in the script we check each circuit in the ceremony e,
- some fields that are required but should not be (i.e., `circuitInputs`).

File: packages/actions/src/helpers/tasks.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            taskArgs.solidityVersion,
+            taskArgs.wasmPath,
+            taskArgs.circuitInputs,
+            taskArgs.verifierTemplatePath,
+            deployer,
+            taskArgs.coordinatorUID
+        )
+
+        // if we are here it is because it didn't throw so we can safely assume that it all was veriifer
+        console.log(`[+] The artifacts generated by the ceremony ${taskArgs.ceremonyPrefix} are valid`)
```
Comment (0xjei) [inline packages/actions/src/helpers/tasks.ts:35]: Correct, +1

File: packages/actions/src/types/index.ts [hunk 4]
```diff
@@ -480,6 +480,34 @@ export type ParticipantDocumentReferenceAndData = {
...
+ * @property {string} directoryRoot - the root directory of the ceremony.
+ * @property {string} potLocalFilePath - the local path of the pot file.
+ * @property {string} r1csLocalFilePath - the local path of the r1cs file.
+ * @property {string} finalZkeyLocalFilePath - the local path of the final zKey file.
+ * @property {string} lastZkeyLocalFilePath - the local path of the last zKey file.
+ * @property {string} verifierLocalFilePath - the local path of the verifier file.
+ * @property {string} verificationKeyLocalFilePath - the local path of the verification key file.
+ * @dev must be used for generating fake/mock documents when testing.
+ */
+export type CeremonyArtifacts = {
```
Comment (0xjei) [inline packages/actions/src/types/index.ts:498]: NIT: seems that there are several redundant pieces of information (`circuitId`, `directoryRoot`, `circuitPrefix`), how come they were included in this type (`CeremonyArtifacts`)?

File: packages/actions/src/types/solc.d.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+/** Declaration file generated by dts-gen */
```
Comment (0xjei) [inline packages/actions/src/types/solc.d.ts:1]: Correct!

File: packages/backend/src/functions/ceremony.ts [hunk 6]
```diff
@@ -24,46 +24,59 @@ dotenv.config()
  * @todo this methodology for transitioning a ceremony from `scheduled` to `opened` state will be replaced with one
  * that resolves the issues presented in the issue #192 (https://github.com/quadratic-funding/mpc-phase2-suite/issues/192).
  */
-export const startCeremony = functions.pubsub.schedule(`every 30 minutes`).onRun(async () => {
-    // Get ready to be opened ceremonies.
-    const scheduledCeremoniesQuerySnap = await queryCeremoniesByStateAndDate(CeremonyState.SCHEDULED, true, "<=")
+export const startCeremony = functions
```
Comment (0xjei) [inline packages/backend/src/functions/ceremony.ts:27]: Cool, thx for taking care of upgrading the memory

**Review by 0xjei — CHANGES_REQUESTED** [review]
Nice progress! :+1:  We have almost finished this part :D Only the last finishing touches are missing to close the verification of the finalization of the ceremony, let's goo :fire:

File: packages/actions/src/types/index.ts (hunk 4 above)
Comment (ctrlc03) [reply packages/actions/src/types/index.ts:517]: I believe this info is useful for the verification as you can see in the code, and returning it from the download artifacts functions saves us from code duplications

**Review by 0xjei — APPROVED** [review]
LGTM! Great job! :star:

**Comment by gurrpi** [comment]
`yarn test:dev` passed locally. Please ignore the false failure of CI for now.

```
Test Suites: 15 passed, 15 total
Tests:       4 skipped, 150 passed, 154 total
Snapshots:   0 total
Time:        254.885 s
```

**Pull request merged**